### ISSUE TITLE
Mapping register/packoffset to what elements in TQualifier do?

### ISSUE THREAD

**Issue by dankbaker** [KhronosGroup/glslang#463]
I'm trying to get the explicit bindings in HLSL to work correctly, but I don't understand all the fields in TQualifer. Layout and Set make sense, What happens when conflicting fields are declared? But specifically, these are my guesses, are they right?

1. layoutAlign - is this the alignment in bytes of the current field? Does this need to be filled out?

2. layoutlocation - is this a virtual register number assuming there are 4 component registers in the case of globals?

3. layoutcomponent - the component used in conjunction with layoutlocation? if I have a 2d value starting at y would that be valid?

4. layoutIndex -How is different then layoutlocation? Or is this used for objects where layoutlocation is used for numeric uniforms?

5. layoutOffset - is this a master override for all of the others, an offset in bytes to the desired type? Obv only valid for uniform numerics, not objects.

**Comment by dankbaker** [comment]
Also, looking through the allocation code I don't think it's quite right. Just to be clear about the "register" keyword and the "packoffset" keyword.  packoffset is essentially a superset of register. Register was a dx9 concept since there were individual registers pack then, but in DX10 that wasn't the case. So packoffset added 2 things:

1) sub component marking

2)  could mark things not on global scope (e.g. inside a cbuffer or tbuffer)

**Comment by johnkslang** [comment]
What would be great to have here is the HLSL -> GLSL mapping as far as all the register letters.

The current mapping is a simple placeholder.  What maps to an opaque uniform location?  What maps to a binding?  What maps to a descriptor set (nothing I think), etc.  If you or others know the mapping, I can pretty easily hook everything up correctly.

**Comment by johnkslang** [comment]
The AST layoutAlign/Location/Component/Index/Offset are essentially as defined by both GLSL and SPIR-V.

**Comment by dankbaker** [comment]
Here's the thing, almost none of it maps to anything in D3D12: Because you have to create a remap table when creating a D3D12 pipeline. So all that happens is becomes an index into another table. The only way to make it work would be to allow the app to pass in a remapping table.

The only thing that really maps to anything in D3D12 is register(c<numeric>) (only on the c registers) and packoffset which really do exactly the same thing, just that packoffset is more general where register defaults to the $Globals.

I'm not super familiar with how GLSL defines layout for uniform constants. It seems like there are multiple ways to do something.

In HLSL
```
float2 MyValue : packoffset(c2.y);
```

would be 'register' number 2, y and z components. I would think this would be Location 2, component 1 in GLSL looking through the code? And index, offset, align aren't used? Is that correct? Or would it be the offset in memory which would be 2 * sizeof(vec4) + 1 * sizeof(float) = 36 bytes?

**Comment by johnkslang** [comment]
My question more about what the "c" meant.  There are lots of letters.  E.g., does c2 mean binding 2, or location 2, etc.

**Comment by danginsburg** [comment]
So speaking from D3D11 (and being unfamiliar with D3D12 HLSL), here are some of the uses we would have in our shaders:

- cbuffer bufferName : register( b# ).  Here b# corresponds to layout( binding = # ) uniform bufferName {...}
- Texture2D name : register( t#).  Here t# corresponds to layout( binding = # ) uniform texture2D name;
- packoffset as described by Dan B in cbuffers used to manually layout constant buffers (i.e. we produce packoffset qualifiers on each structure member to match what std140 would produce in GLSL automatically allowing for the same in-memory layouts across APIs)

In addition to this though, I think there is going to be a bunch of reflection information we'll need to get back.  For example:

- cbuffers/textures declared without register will need to be auto assigned by the parser and then queried afterwards.  Most of our HLSL does not actually use registers at the moment, but instead queries D3D.  We are going to need something resembling a subset of ID3D11ShaderReflection: (https://msdn.microsoft.com/en-us/library/windows/desktop/ff476590(v=vs.85).aspx) .  I would need to go through our code to say exactly what subset of that we need.
- I suspect we will need to invent some additional HLSL syntax of some kind to allow things like descriptor sets to be declared in the shader.  Right now there would be no way I can think of to express that in HLSL.
- There are probably subtle difference between the meanings of these register qualifiers in SM30 (D3D9)/SM40-50(D3D10/11)/SM51.  I'm not super familiar with the differences, but you may need to allow the input to tell you which shader model they are compiling for to actually interpret them as intended.  Or, define a limited subset of HLSL shader models that is actually supported.

I will be able to provide more precise feedback when I start integrating this into our shader pipeline.

**Comment by dankbaker** [comment]
Yes, current syntax won't be sufficient because we need a descriptor set index.  D3D11 had different register banks for different types, there is no analog in Vulkan. D3D12 hacked it by using a table to be passed in during pipeline creation. It is pretty clunky.

In terms of what the register actually does, with exception to the c register (which are a dx9 syntax) they more or less reference a "binding" slot. However, there isn't an obvious way to map them to Vulkan because textures, cbuffers, and samplers all came out of different equivalent "descriptor sets".  You can't set them to set 0, but cbuffers, textures, samplers will all collide spots. 

My opinion is to do the following:

1) Treat register of non C values as the equivalent of layout(binding=[numeric]). This will cause descriptor conflicts if additional info isn't specified since set will default to 0. Alternatively, could simply define deafults to be something like, cbuffers = set 0, textures = set 1, sampler = set 2, uavs = set 3.  Thus, register(b1) becomes equivalent to layout(set=0,binding=1) and register(t3) becomes layout(set=1,binding=3). 

2) The use of c registers not in a cbuffer will default to binding=0, set=0, location=[numeric]

3) If there are multiple bindings, apply the last one.

4) Add layout syntax equivalent to how GLSL works, except it would be post-pended, described in issue #454. This syntax is valid in HLSL since the HLSL compiler can ignore them.

I'm more then willing to take a stab at above rules that I think would make sense, but I wasn't exactly sure how to map packoffset into GLSL style, wether it was location/component or an offset.

**Comment by johnkslang** [comment]
> I'm more then willing to take a stab at above rules that I think would make sense, but I wasn't exactly sure how to map packoffset into GLSL style, wether it was location/component or an offset

The methods `HlslParseContext::handlePackOffset()` and `HlslParseContext::handleRegister` are all set up to do this, minus getting a couple details correct (thanks Dan and Dan).  I think I have the whole picture now.

**Comment by dankbaker** [comment]
> The methods HlslParseContext::handlePackOffset() and HlslParseContext::handleRegister are all set up to do this, minus getting a couple details correct (thanks Dan and Dan). I think I have the whole picture now.

Are you against adding the layout syntax as a post expression modifier? It doesn't solve what to do with register naming yet, but it does allow a path to bind things. Our compilation path premarks things for placement, so it's fairly straightforward to do this.

Without adding the layout syntax, it's currently impossible to use D3D12 style HLSL shaders in Vulkan. At least, not without creating a custom SPIRV patcher.

**Comment by johnkslang** [comment]
> Are you against adding the layout syntax as a post expression modifier?

I agree we need to add syntax, in addition to getting the mappings right for what is present.

I have a preference to do GLSL syntax the way GLSL did it, in front.  Then, we have post-decl syntax for HLSL stuff and pre-decl for GLSL, and both being the way they both are in their native languages. This also allows the recursive decent to be more accepting of multiple languages, so I think there are a few things in favor of doing what you want, but as a pre-decl syntax instead of post-decl.

I need more time to digest, but at a higher level, if there is an agreeable mapping between existing HLSL register names and GLSL binding/location/etc., I'd like to implement that, and then use additional syntax for things like descriptor set, as in GLSL.  This is a pretty important area to get lined up to avoid some frustrating turn on issues.

**Comment by danginsburg** [comment]
One concern I have about giving each resource its own descriptor set is that some implementations have very limited number of sets (or poor performance when using many sets).  An alternative to doing that (which is what we currently do) is you could place them all in descriptor set 0 if they don't have this new syntax.  To avoid binding collisions across shader stages and resource types, the user could pass in a starting offset for each resource for each shader stage.

For example, what we do is start UBO bindings at 0=PS, 11=VS, 22=GS, etc.) and sampler bindings start at the end of UBO bindings (i.e. 5*12=60), and then have a content-based per-stage breakdown for samplers.  

If you assign each resource to its own set for each stage that's a lot of descriptor sets.  If you don't do it for each stage, how do you know where to begin the binding for each stage?  

I think there are lots of valid usages, I'm proposing options are:
1. Application provides its own pre-decl for descriptor set and binding with register syntax, and if that's provided that's what you use.
2. If no descriptor set or binding provided, HLSL -> SPIR-V translator assumes descriptor set 0 and gets a starting binding for each shader stage for per resource type as some kind of input to the library.

@dankbaker Would that work for you?

**Comment by dankbaker** [comment]
It doesn't matter much to us, because we will explicitly use the suggested layout(set,binding) syntax to bind everything into the descriptor set we want. We have a shader process tool which allocates the bindings very specifically for where they are supposed to go.  So, we just need some new syntax for explicit layouts (GLSL style is fine) and Oxide's good to go (we'll just #register to do nothing).

However, If you're going to put syntax in for rebasing I would expand that idea a little bit and allow more general register remapping, so, you'd see something like this:
```
MapRegisterRange(ps_5_0 vs_5_0 start=b0, end=b16,set=0,startbinding=0);
MapRegisterRange(ps_5_0 vs_5_0 start=t0, end=t16,set=1,startbinding=0);
MapRegisterRange(ps_5_0 start=t16, end=t32,set=2, startbinding=0);
```

Which would define Constant buffer slots 0 to 16 to be in set 0, texture slots 0 to 16 to be in set 1, texture slots 16 to 32 to be in set 2 (but texture16 would=binding slot 0), and only for pixel shaders. This would mimic the binding table that you would pass into D3D to get the bindings. It's a little bit verbose but would allow register ranges from any shader stage to map into whereever the user wants them.

or conversly
```
MapRegisterRange(vs_5_0 start=t0, end=t16,set=0,, startbinding=0);
MapRegisterRange(ps_5_0 start=t0, end=t16,set=0, startbinding=16);
```
which would map PixelShader t registers to 16-32, and vertexshader t register to 0-16. 

The default (D3D11) behavior would probably be go create a different set for each stage and each type in some cardinal order, but that is quite a large number of descriptor sets as you point out.  We bind uniformly across all the stages. In D3d11, this means setting the same things in multiple stages, but in D3D12 is like Vulkan in that regard.

**Comment by johnkslang** [comment]
> So, we just need some new syntax for explicit layouts (GLSL style is fine) and Oxide's good to go (we'll just #define register to do nothing).

We could set this up that if GLSL style `layout(set=,binding=,offset=,location=,component=)` sets something in a pre-decl, then an HLSL post-decl does not override it.  Key is to have a rule.

**Comment by johnkslang** [comment]
FYI, I'm currently coding up the GLSL `layout()` qualifier syntax for a pretty full set of GLSL layout-qualifier identifiers.  That is, we'll have the pre-decls needed soon.

**Comment by dankbaker** [comment]
BTW, Doing some experimentation, looks like packoffset is equiv of offset with the formula RegisterNumber*16 + 4*component (16 = sizeof(float4) and 4=sizeof(4) ). Is that what others found?

**Comment by johnkslang** [comment]
> BTW, Doing some experimentation, looks like packoffset is equiv of offset with the formula RegisterNumber x 16 + 4 x component (16 = sizeof(float4) and 4=sizeof(4) ). Is that what others found?

That is already be implemented in `handlePackOffset()`, so I don't think is one of the holes, yes?:

```
    ...
    if (location.size() == 0 || location[0] != 'c') {
        error(loc, "expected 'c'", "packoffset", "");
    ...
    type.getQualifier().layoutOffset = 16 * atoi(location.substr(1, location.size()).c_str());
    if (component != nullptr) {
        int componentOffset = 0;
        switch ((*component)[0]) {
        case 'x': componentOffset =  0; break;
        case 'y': componentOffset =  4; break;
        case 'z': componentOffset =  8; break;
        case 'w': componentOffset = 12; break;
   ...
```

**Comment by dankbaker** [comment]
Assuming that offset does what I think it does, yes, that's correct. I was asking more if the GLSL matched the HLSL so I could validate it.

The register(c) layout stuff looked wrong to me though.

**Comment by johnkslang** [comment]
The following is heading toward the repository now (waiting for a better internet connection):

```
layout(...) <normal HLSL declaration> : <normal HLSL post-decls>
```

where the "..." is most normal (per-variable) GLSL layout qualifiers, but especially including:

```
  set=
  binding=
  constant_id=
  location=
  push_constant
```

Open question still: when the HLSL post-decls are encountered, should they override or (as I proposed earlier) be ignored, meaning the `layout(...)` overrides.

**Comment by dankbaker** [comment]
I would make it take precedent at least - but I think ignoring it is fine.  There is no way to use the register stuff to properly place things without some other syntax as we've suggested.

One question I have is what happens with a variable declared at a global scope rather then a cbuffer. Would that be valid? Does it get aggregated into a specific cbuffer that's already placed into a slot or is ist just not allowed?

**Comment by johnkslang** [comment]
The GLSL syntax is in now.

> I would make it take precedent at least - but I think ignoring it is fine.

Sorry, I didn't follow, which is which or why ignoring is consistent with take precedent.

I kind of want GLSL settings, for those present, to override the HLSL settings, so that a working HLSL shader can have the GLSL added for vulkan, and get what vulkan needs, without having to edit the syntax DX needed, or will continue to need when the shader is used there.  Is that okay?

**Comment by dankbaker** [comment]
The rules I mean were:

1) Layout will always take precedent over the register or packoffset keyword.  Since the individual parts of the layout are abitrary (you could chose not to select a binding), we have to chose whether to merge them in those cases, ignore the register keyword, or fail them. Any of the 3 options is fine with me.

2) uniforms still need to be in cbuffers or tbuffers, you can't use the layout to insert them into layout positions in other buffers

Regarding syntax, we can #define them out either way. But, you might consider the syntax could be made to be valid at either pre or post decl pretty easily. That is, easy enough to make

```
layout(set=0, binding=0) Texture2D Foo.
Texture2D Foo : layout(set=0, binding=0);
```

I don't have a very strong opinion. I'd just note that since HLSL always used post decls to do layout, it's a little more HLSL like to put them as post rather then pre. I _highly_ suspect that at some point in the future HLSL will be modified to allow a similar syntax as a post decl, but that could easily be added later if we anted.

**Comment by johnkslang** [comment]
Yes, generally agreed with all your points.

- Trivial to accept `layout()` in both places, but of course when HLSL adds syntax, it will be different, and we'll add that too.
- Will make `layout()` "win" either way, at fine granularity (per settable), so they are freely mixable and meaningful when mixed.
- We are generally weak on validating semantics, somewhat intentionally.  I don't want to (right now) implement a layout() on a simple global meaning it is part of some other cbuffer/tbuffer, and it sounds like that's undesirable, but we also don't yet have validation checks that the shader didn't try.

**Comment by johnkslang** [comment]
@danginsburg, I moved the part about auto-assigning and reflection to https://github.com/KhronosGroup/glslang/issues/499.

(Combing over this discussion to fully process it and close it out.)

**Comment by johnkslang** [comment]
Another pending aspect of this discussion is about `c` registers.  I opened #500 for that.

**Comment by johnkslang** [comment]
Closing this out, as it as largely addressed, with two pending sub-subjects (now in #499 and #500), and specific issues with the current state of things can be new issues.  This one is getting too big to sort through and track what's done and not done.
